### PR TITLE
Request the user upgrade if the Android SDK version is 24

### LINKED ISSUE #7588: Android SDK 24 can lead to crashes with devices and Macs, but flutter tool says 24 is OK
We diagnosed a problem with crashing from device+flutter+mac. After the user upgraded to Android SDK 25, the crashes stopped. See https://github.com/flutter/flutter/issues/7332#issuecomment-274401118

We might want to change `flutter tool` to prompt the user to upgrade to Android 25. Right now, flutter tool says 24 is OK.

### PR DESCRIPTION
**Pull request by johnmccutchan**
- [x] Detect Android SDK version 24 and request the user upgrade.
- [x] Consolidate on one definition of the `Version` class.

Fixes https://github.com/flutter/flutter/issues/7588

### CODE REVIEW

**Comment by johnmccutchan** [comment]
@devoncarew @jakobr-google

File: packages/flutter_tools/lib/src/android/android_sdk.dart [hunk 1]
```diff
@@ -229,6 +229,9 @@ class AndroidSdkVersion implements Comparable<AndroidSdkVersion> {
   String get apksignerPath => getBuildToolsPath('apksigner');
 
   List<String> validateSdkWellFormed({bool requireApkSigner = true}) {
+    if (buildToolsVersion.major == 24) {
```
Comment (mit-mit) [inline packages/flutter_tools/lib/src/android/android_sdk.dart]: Shouldn't this be `<= 24`?

**Review by mit-mit — COMMENTED** [review]
Thanks John, one questions about the version check.

File: packages/flutter_tools/lib/src/android/android_sdk.dart [hunk 2]
```diff
@@ -229,6 +229,9 @@ class AndroidSdkVersion implements Comparable<AndroidSdkVersion> {
...
+      return <String>['Android version 24 is not supported. Please upgrade.'];
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/android/android_sdk.dart]: nit: `Android SDK`?

**Comment by devoncarew** [comment]
lgtm w/ the version check change

**Comment by johnmccutchan** [comment]
I've updated the message and the test to be < minimumAndroidSdkVersion. PTAL.

File: packages/flutter_tools/lib/src/android/android_sdk.dart [hunk 3]
```diff
@@ -229,6 +232,10 @@ class AndroidSdkVersion implements Comparable<AndroidSdkVersion> {
   String get apksignerPath => getBuildToolsPath('apksigner');
 
   List<String> validateSdkWellFormed({bool requireApkSigner = true}) {
+    if (buildToolsVersion.major < minimumAndroidSdkVersion) {
```
Comment (mit-mit) [inline packages/flutter_tools/lib/src/android/android_sdk.dart]: I don't think this is entirely right. The `Android SDK` and `Android SDK Build-Tools` are separate installs, are versioned independently, and can be used in different versions:
https://developer.android.com/studio/command-line/index.html

So AFAIK would be better if we had two different min variables,  `minimumAndroidSdkVersion` and `minimumAndroidBuildToolsVersion`. Initially these can both be set to 25.

**Review by mit-mit — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/lib/src/android/android_sdk.dart (hunk 3 above)
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/android/android_sdk.dart]: If you read the code in android_sdk.dart you'll see that we only select build tools versions that match the major version number of the platform tools. I'm happy to do the version check against both versions but flutter_tools requires platform version and build tools major version match

Comment (mit-mit) [inline packages/flutter_tools/lib/src/android/android_sdk.dart]: Makes sense for build and platform tools to be in sync. But there are AFAIK actually four artefacts in total: SDK Platform, SDK Build Tools, SDK Platform-tools, and SDK Tools:
https://developer.android.com/studio/intro/update.html#sdk-manager

Must they all four be in sync?

I thought not -- for example, prior to https://github.com/flutter/flutter/pull/8461/files our samples were using SDK 25 but build tools 24. 

Or am I totally misunderstanding this?

Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/android/android_sdk.dart]: I thin you are right that there are four artifacts of the Android SDK but I don't think it matters for this CL. Here is a breakdown of how things work today, what this CL does, and what we might want to do in the future:

Today:
Flutter looks at the platform and build tools versions and selects a version of build tools that has the same major version of the platform. The result of this is that flutter guarantees platformVersion.major == buildToolsVersion.major.

Flutter does not look at platform tools or SDK tools today. Maybe it should, I don't know.

This CL:
Detects if the platform/build tool major versions (they are always the same for Flutter right now) are >= 25.

Future:
In the future we may want to validate version numbers of each of these artifacts individually and may eventually need to build a blacklist / whitelist around version combinations.

Comment (jakobr-google) [inline packages/flutter_tools/lib/src/android/android_sdk.dart]: This describes how the legacy build works. We're switching to Gradle-based builds, where flutter tools don't get to select the SDK or build tools versions. Instead, Gradle uses the versions specified in the build.gradle file, so it doesn't really matter what version flutter thinks it has 'selected'.

Gradle will also automatically download the versions of build tools and SDK it needs, so there's no point in checking that they're installed before we call Gradle.

We currently still support the legacy builds, where this check (and the SDK/build tools selection) matters, but once we remove that and switch completely to Gradle (#8455), all we need to check is that the SDK/build tools licenses have been accepted (#8093).

Comment (mit-mit) [inline packages/flutter_tools/lib/src/android/android_sdk.dart]: OK, sounds like we have some later cleanup to do. Logged https://github.com/flutter/flutter/issues/8525 for that.

**Comment by mit-mit** [comment]
Marking this approved now that we have a cleanup bug filed.

**Review by mit-mit — APPROVED** [review]

**Comment by eseidelGoogle** [comment]
This fixes https://github.com/flutter/flutter/issues/7588 which is the only remaining software bug open for our march 8th milestone. :)  If this is needed for March 8th we ideally want to get this in this morning, since @cbracken plans to make our first branch for that release this afternoon.

**Comment by cbracken** [comment]
Does this still need to make it into the `hackathon` branch?

**Comment by eseidelGoogle** [comment]
A this point I think we can just add this to the list of issues to watch out for.

**Pull request merged**